Tribler/community/tunnel/hidden_community.py

```diff
@@ -228,53 +258,103 @@ def monitor_downloads(self, dslist):
+
+    def on_dht_request(self, messages):
+        for message in messages:
+            info_hash = message.payload.info_hash
+
+            @call_on_reactor_thread
+            def dht_callback(info_hash, peers, _):
+                if not peers:
+                    peers = []
+                meta = self.get_meta_message(u'dht-response')
+                circuit_id = message.payload.circuit_id
+                # Send the list of peers for this info_hash back to the requester
+                dht_response_message = meta.impl(distribution=(self.global_time,), payload=(message.payload.circuit_id,
+                                                                                            message.payload.identifier,
+                                                                                            message.payload.info_hash,
+                                                                                            encode(peers)))
+                circuit = self.exit_sockets[circuit_id]
+                circuit.tunnel_data(message.candidate.sock_addr, TUNNEL_PREFIX + dht_response_message.packet)
+
+            self._logger.debug("Doing dht hidden seeders lookup for info_hash %s" % info_hash.encode('HEX'))
+            self.dht_lookup(info_hash, dht_callback)
+
+    def check_dht_response(self, messages):
+        for message in messages:
+            if not self.is_relay(message.payload.circuit_id):
+                request = self.request_cache.get(u"dht-request", message.payload.identifier)
+                if not request:
+                    yield DropMessage(message, "invalid dht-response identifier")
+                    continue
 
-    def do_lookup(self, info_hash):
-        # Get seeders from the DHT and establish an e2e encrypted tunnel to them
-        @call_on_reactor_thread
-        def dht_callback(info_hash, peers, _):
-            if not peers:
-                return
+            yield message
+
+    def on_dht_response(self, messages):
+        for message in messages:
+            self.request_cache.pop(u"dht-request", message.payload.identifier)
+
+            info_hash = message.payload.info_hash
+            _, peers = decode(message.payload.peers)
 
             blacklist = self.dht_blacklist[info_hash]
 
+            if self.notifier:
+                self.notifier.notify(NTFY_TUNNEL, NTFY_DHT_LOOKUP, info_hash.encode('hex')[:6], peers)
+
             # cleanup dht_blacklist
             for i in xrange(len(blacklist) - 1, -1, -1):
                 if time.time() - blacklist[i][0] > 60:
                     blacklist.pop(i)
-
             exclude = [rp[2] for rp in self.my_download_points.values()] + [sock_addr for _, sock_addr in blacklist]
-            for peer in set(peers):
+            for peer in peers:
                 if peer not in exclude:
+                    self.add_discovered_candidate(Candidate((peer[0], peer[1]), tunnel=False))
```

Comment:
Yep! I'll revert this (and the rest of the key-circuit modifications). Your
explaination about the sock-addr that's set as source of the keyrequest
message makes sense. I did not realize the possibility of sending messages
to an exit socket of a circuit, I'll have a look tonight whether I can see
another cause why the messages are not coming through using the original
code.

Thanks for your help!

Op maandag 29 juni 2015 heeft Niels Zeilemaker <notifications@github.com>
het volgende geschreven:

> In Tribler/community/tunnel/hidden_community.py
> <https://github.com/Tribler/tribler/pull/1510#discussion_r33505489>:
>
> >                  if peer not in exclude:
> > +                    self.add_discovered_candidate(Candidate((peer[0], peer[1]), tunnel=False))
>
> That's only due to the change to the key-request right, I would simply
> revert that
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Tribler/tribler/pull/1510/files#r33505489>.
>